### PR TITLE
Implement an interface based architecture rather than concrete classes in the API's

### PR DESCRIPTION
**Pull request by Colin-Stone**
Implemented and interface and abstraction. See api/ZoweTree for the way its broken down. Because of the wide ranging effect of the changes have a agreed to merge after the CLI extensibility issue has been merged

### CODE REVIEW

File: src/jobs/jobNodeActions.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+*                                                                                 *
+*/
+
+import * as vscode from "vscode";
+import { DeleteJobs } from "@brightside/core";
+import * as nls from "vscode-nls";
+import { Job } from "../ZoweJobNode";
+const localize = nls.config({ messageFormat: nls.MessageFormat.file })();
+
+export async function deleteJob(node: Job) {
```
Comment (zFernand0) [inline src/jobs/jobNodeActions.ts]: Should the actions be part of the ZoweJobNode class?
(Just wondering what your thoughts are, no required change)

File: src/uss/ussNodeActions.ts [hunk 2]
```diff
@@ -91,7 +92,7 @@ export async function createUSSNodeDialog(node: ZoweUSSNode, ussFileProvider: US
     }
 }
 
-export async function deleteUSSNode(node: ZoweUSSNode, ussFileProvider: USSTree, filePath: string) {
+export async function deleteUSSNode(node: IZoweTreeNode, ussFileProvider: IZoweTree<IZoweTreeNode>, filePath: string) {
```
Comment (zFernand0) [inline src/uss/ussNodeActions.ts]: Same question here, should these set of actions be part of the corresponding class?

File: src/ZoweNode.ts [hunk 3]
```diff
@@ -25,7 +26,7 @@ const localize = nls.config({ messageFormat: nls.MessageFormat.file })();
  * @class ZoweNode
  * @extends {vscode.TreeItem}
  */
-export class ZoweNode extends vscode.TreeItem implements IZoweTreeNode {
+export class ZoweNode extends ZoweTreeNode implements IZoweDatasetTreeNode {
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply src/ZoweNode.ts]: I've always wanted to do that but was unsure of any consequences.  I will take this as a go ahead an do it recommendation

File: src/api/ZoweTree.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+/**
+ * The base interface for Zowe tree browsers that implement the
+ * vscode.TreeDataProvider.
+ *
+ * @export
+ * @interface IZoweTree
+ * @extends {vscode.TreeDataProvider<T>}
+ * @template T provide a subtype of vscode.TreeItem
+ */
+export interface IZoweTree<T> extends vscode.TreeDataProvider<T> {
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply src/api/ZoweTree.ts]: That's me being all softly-softly and not wishing to alter files that others had contributed. I totally agree with you and now feel empowered

File: src/jobs/jobNodeActions.ts (hunk 1 above)
Comment (Colin-Stone) [reply src/jobs/jobNodeActions.ts]: Agree 100%.

File: src/uss/ussNodeActions.ts (hunk 2 above)
Comment (Colin-Stone) [reply src/uss/ussNodeActions.ts]: Yes did the same. Should probably look to do rename also, but leaving until merging after cli extensions PR has gone in

File: src/ZoweNode.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (zFernand0) [reply src/ZoweNode.ts]: The only thing we need to keep in mind in the future is if we change any extensible API names. Everything else should be fare game. 😄

File: src/api/ZoweTree.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (zFernand0) [reply src/api/ZoweTree.ts]: I think that's good enough.
As they say: `If it isn't broken, don't try to fix it` 😄

File: src/ZosJobsProvider.ts [hunk 5]
```diff
@@ -339,13 +244,66 @@ export class ZosJobsProvider implements IZoweTree<Job> {
...
     }
 
+    /**
+     * Change the state of an expandable node
+     * @param provider the tree view provider
+     * @param element the node being flipped
+     * @param isOpen the intended state of the the tree view provider, true or false
+     * N.B. Specific implementation for Jobs as owner field set.
+     */
+    public async flipState(element: IZoweJobTreeNode, isOpen: boolean = false) {
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply src/ZosJobsProvider.ts]: I had intended to do this but obviously missed the deleting of the old once I had set up the single implementation

**Comment by zFernand0** [comment]
This PR should be good to merge after #440.
Also, please ignore codecov misleading errors. See #473

File: i18n/sample/src/USSTree.i18n.json [hunk 6]
```diff
@@ -2,6 +2,7 @@
     "Favorites": "Favorites",
     "ussTree.error": "Error encountered in ",
     "ussFilterPrompt.log.debug.promptUSSPath": "Prompting the user for a USS path",
+    "ussTree.error": "Error encountered in ",
```
Comment (zFernand0) [inline i18n/sample/src/USSTree.i18n.json]: Is it a problem that this error message appears twice?

**Pull request merged**